openedx_external_enrollments/external_enrollments/icc_external_enrollment.py

```diff
@@ -0,0 +1,198 @@
+        """
+
+        enrollment_data = {}
+        try:
+            icc_user = self._get_icc_user(data)
+            enrollment_data = {
+                "wstoken": settings.ICC_API_TOKEN,
+                "wsfunction": settings.ICC_ENROLLMENT_API_FUNCTION,
+                "enrolments[0][roleid]": settings.ICC_LEARNER_ROLE_ID,
+                "enrolments[0][userid]": icc_user.get('id'),
+                "enrolments[0][courseid]": course_settings.get("external_course_run_id"),
+            }
+        except Exception:
+            pass
+
+        return enrollment_data
+
+    def _get_enrollment_url(self, course_settings):
+        """
+        Method that returns base url for the ICC integration API.
+        """
+
+        return settings.ICC_BASE_URL
+
+    def _get_icc_user(self, data):
+        """
+        Method that look for the user in ICC database through the API, if the user is not
+        found then it calls create method.
+        """
+        log_details = {
+            "url": settings.ICC_API_TOKEN,
+        }
+        icc_user = {}
+
+        try:
+            _data = {
+                "wstoken": settings.ICC_API_TOKEN,
+                "wsfunction": settings.ICC_GET_USER_API_FUNCTION,
+                "criteria[0][key]": "email",
+                "criteria[0][value]": data.get('user_email'),
+            }
+            response = requests.post(
+                url=settings.ICC_BASE_URL,
+                data=_data,
+            )
+            log_details['request_payload'] = _data
+
+        except Exception as error:
+            LOG.error('Failed to retrieve ICC user. Reason: %s', str(error))
+            log_details['request_payload']['wstoken'] = 'icc-api-token'
+            log_details['response'] = {'error': 'Failed to retrieve ICC user. Reason: ' + str(error)}
+            EnrollmentRequestLog.objects.create(  # pylint: disable=no-member
+                request_type=str(self),
+                details=log_details,
+            )
+            pass
+
+        else:
+            icc_user = self._get_icc_user_from_xml_response(response, 'get_user')
+
+            # Ensure the user creation avoiding failure for username duplicates
+            duplicated_username = False
+            call_counter = 0
+            while not icc_user and call_counter <= self.USER_CREATION_MAX_RETRIES:
+                icc_user = self._create_icc_user(data, duplicated_username)
+                duplicated_username = True
+                call_counter = call_counter + 1
+
+        return icc_user
+
+    def _create_icc_user(self, data, duplicated_username):
+        """
+        Method that creates a user in the ICC database based int the  current user logged in data.
+        Also takes care of duplicated usernames.
+        """
+        log_details = {
+            "url": settings.ICC_CREATE_USER_API_FUNCTION,
+        }
+        icc_user = {}
+
+        try:
+            user, _ = get_user(email=data.get('user_email'))
```

Comment:
Please assign a default value to the .get call unless the default should be None. @ivanvgh